File: Very_basic_examples/switch-case1.rs
// Assembly code: https://rust.godbolt.org/z/dzWjqPfW9
const PROGLEN: usize = 0x10_000;

fn interpret_example(opcodes: &[u8; PROGLEN]) -> i32 {
    let mut instrptr = 0;
    let mut reg: i32 = 0;

    loop {
        instrptr %= PROGLEN;
        match opcodes[instrptr] & 0x07 {
            0x00 => reg += 133,
            0x01 => reg -= 133,
            0x02 => reg += 155,
            0x03 => reg -= 155,
            0x04 => reg += 177,
            0x05 => reg -= 177,
            0x06 => instrptr = reg as usize, // JMP
            0x07 => return reg,
            _ => (),                         // this syntax needs to 0x07 (Rust 1.70)
        }
        instrptr += 1;
    }
}

File: Very_basic_examples/switch-case1.c
// Assembly code: https://c.godbolt.org/z/K46bK1rbK
#include <stddef.h>
#include <stdint.h>

#define PROGLEN 0x10000

int32_t interpret_example(unsigned char opcodes[PROGLEN]) {
    size_t instrptr=0;
    int32_t reg=0;

    while (1) {
        instrptr %= PROGLEN;
        switch (opcodes[instrptr++] & 0x07) {
            case 0x00: reg += 133; break;
            case 0x01: reg -= 133; break;
            case 0x02: reg += 155; break;
            case 0x03: reg -= 155; break;
            case 0x04: reg += 177; break;
            case 0x05: reg -= 177; break;
            case 0x06: instrptr = reg; break; // JMP
            case 0x07: return reg; break;
        }
    }
}
